transcendental equation solver for solveset: transolve

Yathartha22 opened the pull request:
This PR aims at extending capability of `solveset` by implementing `transolve` that is a helper to solveset for solving transcendental equations.

The following things are done:

- [x] API design and flow of transolve
- [x] Exponential solver
- [x] import test from solve
- [x] Fix pep8
- [x] address comments

The following needs to be done:

* solving exponential equation in complex domain~ (will be solved later, added tests as XFAIL as of now)

- [x] XFails

- [x] Documentation with proof of correctness.

this is an alternative to #13045.
~#10864~

ping @aktech @smichr

Yathartha22 committed:
initial commit for transolved and exponential solver

Yathartha22 committed:
added tests for exponential solver

Yathartha22 committed:
added docstring for exponential solver

Yathartha22 committed:
Merge branch 'master' of https://github.com/sympy/sympy into transolve_gsoc

Yathartha22 commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1749,58 @@ def test_issue_14454():
+    f2 = x**x
+
+    assert solveset(e1, x, S.Reals) == FiniteSet(-3*log(2)/(-2*log(3) + log(2)))
+    assert simplify(solveset(e2, x, S.Reals)) == FiniteSet(4/S(15))
```
Need to return a simplified result for such equations.
eg: 2**x - 32 currently gives `{log(32)/log(2)}`, should rather be {5}.
Is it a good idea to use `simplify()` [here](https://github.com/sympy/sympy/blob/master/sympy/solvers/solveset.py#L930)?

asmeurer replied:
Raw `simplify` can be a very expensive function call and is generally avoided in library code. In the old `solve` there is a flag to enable or disable `simplify`. 

Is it possible instead to invert powers using a two argument log, like `log(32, 2)`?

Yathartha22 replied:
@asmeurer it seems that two argument log does not work well with `imageset`. I tried

```
>>> g = FiniteSet(32)
>>> imageset(Lambda(n, log(n, 2)), g)
{log(32)/log(2)}
```

Yathartha22 replied:
opened #14746 for this. Probably should solve this problem.

asmeurer commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1749,58 @@ def test_issue_14454():
+
+    assert solveset(z**x - y, x, S.Reals) == Intersection(S.Reals, FiniteSet(log(y)/log(z)))
+    assert solveset(y - a*x**b, x) == FiniteSet((y/a)**(1/b))
+    assert solveset(Poly(exp(x) + exp(-x) - 4)) == FiniteSet(log(-sqrt(3) + 2), log(sqrt(3) + 2))
```
Why is `Poly` used here?

Yathartha22 replied:
I think Poly should be removed as there is no condition in solveset that would extract expression from Poly as it does in solve. I might have missed it.
Can we have a check of such case just to increase the coverage?

```
if isinstance(f, Poly):
            f = f.as_expr()
```

asmeurer commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,88 @@ def _solveset(f, symbol, domain, _check=False):
+    rhs = -b
+
+    lhs = expand_log(log(lhs), force=True)
+    rhs = expand_log(log(rhs), force=True)
```
force=True does operations that are not mathematically correct for all values. Is it possible for this function to return wrong results because of this?

smichr replied:
Yes, log(x*y) will expand into log(x)+log(y) but that is not always true. e.g. if x == y == -1, the former is 0 while the latter is `2*pi*I` in SymPy.

aktech commented:
@Yathartha22 This PR doesn't fulfill the minimum requirements for review, which we discussed in the first meeting. I don't see this [comment](https://github.com/sympy/sympy/pull/13045#discussion_r176921547) being taken care of.

Yathartha22 committed:
added documentation for transolve

Yathartha22 committed:
minor changes in code and tests.

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,180 @@ def _solveset(f, symbol, domain, _check=False):
+            else:
+                result = transolve(f, symbol, domain, **flags)
+
+        elif lhs.is_Pow:
```
Not sure about this case. Only one case as of now will be handled here `x**x`. If it seems better we can put a case for this in inver_real.

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,180 @@ def _solveset(f, symbol, domain, _check=False):
+        rhs = rhs_s.args[0]
+        # do we need a loop for rhs_s?
+        # Can't determine a case as of now.
+        if lhs == symbol:
```
This case is added so that even if one tries to solve the equation by calling `transolve`, result can be obtained.
cases like: `2**x  -32` will be handled here.

smichr replied:
yes

Yathartha22 commented:
ping @aktech

Yathartha22 committed:
minor changes

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,181 @@ def _solveset(f, symbol, domain, _check=False):
+       transolve's API and its flow is simple to understand. Unlike _tsolve which
+       computes by solving with lots of recursive calls.
+       transolve eases the task by reducing it to a two step procedure as
+       dicussed below.
```
typos

`inver_real` - `invert_real`
eqaution - equation
generalised - generalized (I know, variant) :-)
dicussed - discussed
doamin - domain
eqquation - equation

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,181 @@ def _solveset(f, symbol, domain, _check=False):
+    {-log(375)/(-log(5) + 2*log(3))}
+    """
+
+    if 'tsolve_saw' not in flags:
```
`flags.setdefault('tsolve_saw', [])`

smichr replied:
Or all in one:
```
if f in flags.setdefault('tsolve_saw', []):
    return...
flags['tsolve_saw'].append(f)
```

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,181 @@ def _solveset(f, symbol, domain, _check=False):
+    else:
+        flags['tsolve_saw'].append(f)
+
+    inverter = invert_real if domain.is_subset(S.Reals) else invert_complex
```
A call to `invert_complex(..., domain)` will handle which inverter to call (like calling to solveset will just work when you pass a domain).

Yathartha22 replied:
Yeah, we can do that. Will add a comment so as to avoid confusion.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,181 @@ def _solveset(f, symbol, domain, _check=False):
+            result = rhs_s
+
+        elif lhs.is_Add:
+            equation = factor(powdenest(lhs-rhs))
```
you have to capture denominators containing possible singularities before doing `lhs-rhs` or else terms with denominators may cancel, e.g. `x*y - 1/x = x*z - 1/x` for which x = 0 is not a solution.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,181 @@ def _solveset(f, symbol, domain, _check=False):
+        result = S.EmptySet
+    else:
+        rhs = rhs_s.args[0]
+        # do we need a loop for rhs_s?
```
```
assert (rhs_s.args) == 1  # can't think of a case where there would be more than one
```

aktech replied:
+1

@Yathartha22 This comment doesn't seem to be addressed as of yet. Never assume anything, always handle all the cases.

Yathartha22 replied:
I will add a loop here rather an `assert` statement.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,181 @@ def _solveset(f, symbol, domain, _check=False):
+            result = rhs_s
+
+        elif lhs.is_Add:
+            equation = factor(powdenest(lhs-rhs))
```
Would `factor_terms` be sufficient for what you are expecting? See its options, too, for handling radicals.

Yathartha22 replied:
I will have a look over it. But here I am trying to create a P.O.S of the equation.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,181 @@ def _solveset(f, symbol, domain, _check=False):
+
+    ii) Invoking the respective helper to solve the equation.
+
+        Once identified what family the equation belongs, respective
```
Once the family type of the equation has been identified, the corresponding helper is called...

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,181 @@ def _solveset(f, symbol, domain, _check=False):
+    Helper function for solving exponential equations.
+
+    This function solves exponential equation by using logarithms.
+    Exponents are converted to a more general form of logs which can further be
```
...which can be solved by solveset.

Yathartha22 committed:
improved documentation and imported tests from solve.

Yathartha22 committed:
minor documentation fixes

Yathartha22 committed:
added proof of correctness and minor changes in documentation

Yathartha22 commented:
@asmeurer @aktech @smichr I think it is ready to be reviewed

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+
+        elif lhs.is_Add:
+            # trying to convert to P.O.S form
+            equation = factor(powdenest(lhs-rhs))
```
space around operator `-`

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+
+    if f in flags.setdefault('tsolve_saw', []):
+        return ConditionSet(symbol, Eq(f, 0), domain)
+    flags['tsolve_saw'].append(f)
```
I don't see any significance of this parameter as of now, neither from the code nor from the tests.

The following would work as well:

```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index b1e24fe..78ea5bd 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -1066,7 +1066,7 @@ def _check_expo(f, symbol):
     return any(_ispow(i) for i in (ad, bd))


-def transolve(f, symbol, domain, **flags):
+def transolve(f, symbol, domain):
     """
     Function to solve transcendental equations. It is a helper to
     solveset and should be used internally as of now. `transolve`
@@ -1254,10 +1254,6 @@ class otherwise `False`. Solving helpers needs to be implemented with
     {-log(375)/(-log(5) + 2*log(3))}
     """

-    if f in flags.setdefault('tsolve_saw', []):
-        return ConditionSet(symbol, Eq(f, 0), domain)
-    flags['tsolve_saw'].append(f)
-
     # invert_complex handles the call to the desired inverter based
     # on the domain specified.
     lhs, rhs_s = invert_complex(f, 0, symbol, domain)
@@ -1286,7 +1282,7 @@ class otherwise `False`. Solving helpers needs to be implemented with
                 result = _solveset(new_f, symbol, domain)

             else:
-                result = transolve(f, symbol, domain, **flags)
+                result = ConditionSet(symbol, Eq(f, 0), domain)

         elif lhs.is_Pow:
             new_f = _expo_solver(lhs - rhs)
```

Yathartha22 replied:
We can remove the condition completely still it will return `ConditionSet` as result is initialised to `ConditionSet`.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+    return (lhs - rhs)
+
+
+def _check_expo(f, symbol):
```
Add tests for this function

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+    Exponential equations are the type of equations which includes a
+    variable located in the exponent.
+    For example `5**(2*x + 3) - 5**(3*x - 1)`.
+    Equations having two arguments (like the one above) or that can be
```
What do you mean by `Equations having two arguments `, define.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+    For example `5**(2*x + 3) - 5**(3*x - 1)`.
+    Equations having two arguments (like the one above) or that can be
+    reduced to two argument form can be handled by this helper, other
+    variants require numerical methods to evaluate.
```
There is no need to mention that, that's an obvious fact which we are not interested in.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+    equation to log form (if possible) and returns the modified
+    equation.
+
+    * Proof of corectness of the method
```
correctness

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+    lhs = expand_log(log(lhs))
+    rhs = expand_log(log(rhs))
+
+    return (lhs - rhs)
```
redundant parenthesis around

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+    This function simply determines whether any of the symbol dependent
+    term contains power. If so there is a chance that the equation is of
+    exponential type, hence returns True otherwise False. It relies on
+    the fact that the exponential equations are of two argument form as
```
define what do you mean by two argument form properly

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+    False
+    """
+    try:
+        a, b = ordered(f.args)
```
please use meaningful variable names.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,319 @@ def _solveset(f, symbol, domain, _check=False):
+        return False
+
+    ad = a.as_independent(symbol)[1]
+    bd = b.as_independent(symbol)[1]
```
same here

aktech reviewed (commented):
Please import all the tests from _tsolve and XFAIL as of now.

Yathartha22 commented:
> Please import all the tests from _tsolve and XFAIL as of now.

I have imported only exponents tests, should I also include tests of other type equations (log, lambert etc.).

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,291 @@ def _solveset(f, symbol, domain, _check=False):
+    return any(_ispow(i) for i in (ad, bd))
+
+
+def transolve(f, symbol, domain, **flags):
```
Do you mean making it `_transolve` ??

aktech replied:
yes

aktech commented:
> I have imported only exponents tests, should I also include tests of other type equations (log, lambert etc.).

Yes, all of them.

Yathartha22 committed:
imported tests for log and lambert form solve, minor documentation fixes,
removed use flag

Yathartha22 committed:
improvements for XFAILS to pass

Yathartha22 committed:
updated and improved documentation for sphinx

Yathartha22 commented:
ping @aktech

Yathartha22 committed:
removed repeated test

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -232,18 +234,22 @@ def _invert_real(f, g_ys, symbol):
+                    #special case: 0**x - 1
+                    return (expo, FiniteSet(0))
+                return (expo, S.EmptySet)
+            except TypeError:
```
Where do you expect this?

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -232,18 +234,22 @@ def _invert_real(f, g_ys, symbol):
+                    #special case: 0**x - 1
+                    return (expo, FiniteSet(0))
+                return (expo, S.EmptySet)
+            except TypeError:
```
Where do you expect this?

Yathartha22 replied:
For equations having symbols, it turns out that relational comparison of symbols with `0` raises an exception. I have added tests in `test_expo_symbols`.

aktech replied:
then in that case you should rather check before comparing, instead of raising a TypeError, whose significance  is hard to find out from the code

aktech replied:
It will catch all kinds of TypeError in that block of code and you don't want that. You should only catch the exact exception only where its known to be expected or check before comparing.

Yathartha22 replied:
I think such comparison should not be made when it comes to sympy. I would rather change 
`base > S.Zero` to `base.is_positive`.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,354 @@ def _solveset(f, symbol, domain, _check=False):
+    `\_tsolve`. The way equations are treated in solve is complicated
+    which make it a bit time consuming. `solve` tries every possibility
+    to solve the equation and in process it gets into each and every
+    condition and starts solving the equation to get the result,
```
`The way equations are treated to solve is complicated` -> This doesn't really provide any useful information.


You should rephrase the following:
```
`solve` tries every possibility
+    to solve the equation and in process it gets into each and every
+    condition and starts solving the equation to get the result,
```

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,354 @@ def _solveset(f, symbol, domain, _check=False):
+    condition and starts solving the equation to get the result,
+    if the result is not achieved the equation is solved by another
+    way. Therefore this series of solving makes it a bit slow.
+    Whereas in `transolve` computation begins only when the equation
```
> replace all the reference of transolve with actual function name, i.e. _transolve

and same for all other instances.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,354 @@ def _solveset(f, symbol, domain, _check=False):
+def _transolve(f, symbol, domain):
+    r"""
+    Function to solve transcendental equations. It is a helper to
+    solveset and should be used internally as of now. `transolve`
```
>replace all the reference of transolve with actual function name, i.e. _transolve

aktech replied:
backticks around `solveset`

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,354 @@ def _solveset(f, symbol, domain, _check=False):
+    How to use `transolve`
+    ======================
+
+    `transolve` should be used as an internal function. There should be
```
> replace all the reference of transolve with actual function name, i.e. _transolve

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,354 @@ def _solveset(f, symbol, domain, _check=False):
+    {-log(375)/(-log(5) + 2*log(3))}
+
+
+    How to use `transolve`
```
> replace all the reference of transolve with actual function name, i.e. _transolve

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,354 @@ def _solveset(f, symbol, domain, _check=False):
+    instances.
+
+
+    How `transolve` works
```
> replace all the reference of transolve with actual function name, i.e. _transolve

Yathartha22 committed:
minor documentation fixes

Yathartha22 commented:
Have a look now @aktech

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1886,7 +1886,6 @@ def test_uselogcombine_1():
     assert solveset_real(log(x + 3) + log(1 + 3/x) - 3) == FiniteSet(
         -3 + sqrt(-12 + exp(3))*exp(S(3)/2)/2 + exp(3)/2,
         -sqrt(-12 + exp(3))*exp(S(3)/2)/2 - 3 + exp(3)/2)
-    assert solveset_real(log(x-3) + log(x+3), x) == FiniteSet(sqrt(10))
```
Why is this test removed?

Yathartha22 replied:
I haven't removed the test. I have moved the test (both the functions: `uselogcombine_1`, `uselogcombine_2`) at last.

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,163 @@ def test_issue_14454():
+    e4 = exp(log(5)*x) - 2**x
+    e5 = exp(x/y)*exp(-z/y) - 2
+    e6 = 5**(x/2) - 2**(x/3)
+    e7 = 2**(x) + 4**(x) + 8**(x) - 84
```
parenthesis around x is redundant

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,163 @@ def test_issue_14454():
+    e5 = exp(x/y)*exp(-z/y) - 2
+    e6 = 5**(x/2) - 2**(x/3)
+    e7 = 2**(x) + 4**(x) + 8**(x) - 84
+    e8 = 4**(x+1) + 4**(x+2) + 4**(x-1) - 3**(x+2) - 3**(x+3)
```
space around operator, for e.g.: `(x+1)` -> (x + 1)

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,163 @@ def test_issue_14454():
+
+    assert _transolve(2**x - 32, x, S.Reals) == FiniteSet(5)
+    assert _transolve(3**x, x, S.Reals) == S.EmptySet
+    assert _transolve(3**x - 9**(x+5), x, S.Reals) == FiniteSet(-10)
```
space around operator

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,163 @@ def test_issue_14454():
+    assert solveset(e8, x, S.Reals) == FiniteSet(2)
+    assert solveset(e9, x, S.Reals) == FiniteSet(log(9*exp(4)/4)/5)
+
+
```
too many lines, remove redundant line

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,163 @@ def test_issue_14454():
+    assert solveset(e9, x, S.Reals) == FiniteSet(log(9*exp(4)/4)/5)
+
+
+    assert solveset_real(-9*exp(-2*x + 5) + 2**(x+1), x) == \
```
space around operator

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,163 @@ def test_issue_14454():
+        imageset(Lambda(n, 3*n*I*pi/log(2)), S.Integers)
+    assert solveset(2**x + 32, x) == \
+        imageset(Lambda(n, (I*(2*n*pi + pi) + 5*log(2))/log(2)), S.Integers)
+    assert solveset(2**(x) + 4**(x) + 8**(x), x) == \
```
redundant parenthesis around `x`

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,163 @@ def test_issue_14454():
+    assert solveset(2**(x) + 4**(x) + 8**(x), x) == \
+        Union(imageset(Lambda(n, I*(2*n*pi - 2*pi/3)/log(2)), S.Integers),
+              imageset(Lambda(n, I*(2*n*pi + 2*pi/3)/log(2)), S.Integers))
+    assert solveset(4**(x+1) + 4**(x+2) + 4**(x-1) - 3**(x+2) - 3**(x+3), x) == \
```
space around operator

aktech reviewed (changes requested):
Please use a plugin for pep8, I have pointed out a few issues, there are a lot of such issues.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,349 @@ def _solveset(f, symbol, domain, _check=False):
+    # on the domain specified.
+    lhs, rhs_s = invert_complex(f, 0, symbol, domain)
+
+    result = ConditionSet(symbol, Eq(f, 0), domain)
```
You can also write this as:
```
unsolved_result = ConditionSet(symbol, Eq(f, 0), domain)
result = unsolved_result
```
and use the `unsolved_result` in Line 1330, rather than duplicating the same thing.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,349 @@ def _solveset(f, symbol, domain, _check=False):
+
+def _check_expo(f, symbol):
+    r"""
+    Helper to check whether an equation is exponential or not.
```
This line just says that it checks if its exponential or not but the description below says it checks only two exponent terms, this is not consistent and misleading.

aktech replied:
Also why does it only looks for exponential with two exponent terms? In `_transolve`'s documentation you have written it can solve Exponential equations which is not correct, based on this.

Yathartha22 replied:
> Also why does it only looks for exponential with two exponent terms? 

Exponents with one term is handled by `_invert_real`, more than two terms are not supported (unless they can be reduced like `4**(x + 1) + 4**(x + 2) + 4**(x - 1) - 3**(x + 2) - 3**(x + 3)`, this will be factored and handled in `is_Mul` in `_transolve`), therfore `_check_expo` only looks for two terms.
Should I make for n terms??

aktech replied:
Which one do you think is better and why and how would you convey this information to a developer? Think about these things.

Yathartha22 replied:
I have made changes in `_check_expo` to work for n arguments.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,349 @@ def _solveset(f, symbol, domain, _check=False):
+    type exponential equations. Two argument type exponential
+    equations are the ones which consists of only two exponent terms.
+    Therefore it returns False if `f` does not contains exactly two
+    arguments.
```
some examples would be good.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,349 @@ def _solveset(f, symbol, domain, _check=False):
+
+        elif lhs.is_Add:
+            # trying to convert to P.O.S form
+            equation = factor(powdenest(lhs - rhs))
```
Use a better variable name, what kind of equation do you expect here?

Yathartha22 replied:
> what kind of equation do you expect here?

Equations that can be factored to a simpler form (possibly product form), equation like `4**(5 - 9*x) - 8**(2 - x)` can be converted to `64*2**(-3*x)*(-1 + 16*2**(-15*x))` by factoring, which is easily handled by `solveset`, or any type of equation that can be changed to a product form.

Better name could be `simplified_equation` or `factored_equation`??

aktech replied:
I understand what that does, think from the perspective of a dev who is reading is for the first time.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,349 @@ def _solveset(f, symbol, domain, _check=False):
+            # check if it is exponential type equation
+            elif _check_expo(equation, symbol):
+                new_f = _expo_solver(equation)
+                result = _solveset(new_f, symbol, domain)
```
what if `new_f` is same as given `f`?

Yathartha22 replied:
I don't think this will happen, because even if an equation gets into the condition, `new_f` will contain expression in `log`. which will be different than `f`.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,349 @@ def _solveset(f, symbol, domain, _check=False):
+
+        elif lhs.is_Pow:
+            new_f = _expo_solver(lhs - rhs)
+            result = _solveset(new_f, symbol, domain)
```
same as [this](https://github.com/sympy/sympy/pull/14736/files#r194921893)

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +989,349 @@ def _solveset(f, symbol, domain, _check=False):
+        # there is chance that ConditionSet is returned from
+        # any of the helper which will have modified equation
+        # therefore to get the original equation in the ConditionSet
+        # this condition is necessary.
```
You don't have to write such comments if you write self-commenting code, as mentioned [here](https://github.com/sympy/sympy/pull/14736/files#r194917194). ;)

Yathartha22 committed:
documentation fxes according to pep8; made `_check_expo` to work for n arguments; used loop in `_transolve` to handle every case;

Yathartha22 commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,164 @@ def test_issue_14454():
+
+@XFAIL
+def test_issue_10864():
+    assert solveset(x**(y*z) - x, x, S.Reals) == FiniteSet(1)
```
this test was previously solved as by `_expo_solver` because `_check_expo` returned True, since then it checked if the equation had power terms, but now `_check_expo` is improved and it checks for `symbol` in the exponent then only returns True.

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,353 @@ def _solveset(f, symbol, domain, _check=False):
+    try:
+        arg1, arg2 = ordered(f.args)
+    except ValueError:
+        return None
```
`None` is returned when more than two arguments are present in the equation.

smichr replied:
But definitely do the test before preparing the args, otherwise anything in the `ordered(f.args)` might fail.

Yathartha22 replied:
Sorry, I missed this.

Yathartha22 replied:
I have added a test for this as well.

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,353 @@ def _solveset(f, symbol, domain, _check=False):
 
+def _expo_solver(f):
+    r"""
+    Helper function for solving exponential equations.
```
I will make it: `Helper function for solving (supported) exponential equations.`

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,353 @@ def _solveset(f, symbol, domain, _check=False):
+    r"""
+    Helper to check whether an equation is exponential or not.
+
+    It filters out power and `exp` terms from the equation and
```
Please explain what do you mean by `exp` and also power. One of them is written as a english word (power) and the other as code (`exp`).

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,353 @@ def _solveset(f, symbol, domain, _check=False):
+    pow_args = f.atoms(Pow, exp)
+    number_of_terms = len(pow_args)
+
+    for i in pow_args:
```
Please use meaningful variable names, `i` doesn't helps.

Yathartha22 replied:
Sorry!

smichr replied:
On the other hand, this is a pretty oft used idiom in SymPy: using `i` when traversing a list. You could also use the underscore to further indicate, "this is just the loop variable and if you want to know what list it belongs to, look at the beginning of the loop."

aktech replied:
I think it's quite handy to see its significance, right inside the loop, instead of jumping to the start of the loop.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,353 @@ def _solveset(f, symbol, domain, _check=False):
+    number_of_terms = len(pow_args)
+
+    for i in pow_args:
+        i_expo = i.args[0] if isinstance(i, exp) else i.exp
```
I don't see any significance for this condition, at least from the tests.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,353 @@ def _solveset(f, symbol, domain, _check=False):
+    number_of_terms = len(pow_args)
+
+    for i in pow_args:
+        i_expo = i.args[0] if isinstance(i, exp) else i.exp
```
Even the following will work, based on current tests, if you think there is any significance of the above condition, then please add it to the tests.

```python
    pow_type_atomic_objects = f.atoms(Pow, exp)

    for atomic_object in pow_type_atomic_objects:
        exponent = atomic_object.exp
        if exponent.has(symbol):
            return True
    return False
```

smichr replied:
I prefer more affirmation of what it is that you want: you need and Add with a term having a factor that is a power with a symbol-dependent exponent. In case your calling code is not that specific, you might want to use:

```
>>> def go(f):
...   for a in Add.make_args(f):
...     for m in Mul.make_args(a):
...         if isinstance(m, (exp, Pow)) and symbol in m.free_symbols:
...             return True
...   return False
... 
>>> go(cos(2**x))
False
```

The problem with asking for atoms is that they come from anywhere.

aktech replied:
Good point.

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1737,164 @@ def test_issue_14454():
+def test_check_expo():
+    x, y, z = symbols('x y z')
+
+    assert _check_expo(3**x - 2, x) == True
```
`==` -> `is`

e.g: `assert _check_expo(3**x - 2, x) is True`

aktech replied:
same for all the similar ones.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,353 @@ def _solveset(f, symbol, domain, _check=False):
+    r"""
+    Helper function for solving exponential equations.
+
+    Exponential equations are the type of equations which has a
```
"A sum of (currently) at most 2 terms with one or both of them having a power with a symbol-dependent exponent."

(That `exp` is a function is an internal detail that I wouldn't bother differentiating between in the docstring, but would make the second example below involve `exp`, e.g. change the `8^` to `exp`.

Yathartha22 committed:
documentation fixes for: `transolve`, `check_expo` and `expo_solver`; improved logic for `check_expo`; improved code quality for `transolve`.

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,365 @@ def _solveset(f, symbol, domain, _check=False):
+
+    for equation in [lhs - rhs for rhs in rhs_s]:
+        if equation == f:
+            if lhs.is_Add:
```
I have created a helper for this case as well as for below case (I don't think there could be more), so that these portion of `transolve` does not get affected when adding new solvers or fixing issues. All the modifications is now restricted to the two internal functions above.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1040,50 +1038,52 @@ def _expo_solver(f):
+    for add_arg in Add.make_args(f):
+        for mul_arg in Mul.make_args(add_arg):
+            if isinstance(mul_arg, (Pow, exp)) and \
+                    mul_arg.exp.has(symbol):
```
`foo.has(symbol)` is a weaker assertion than `symbol in foo.free_symbols`; an expression can have `symbol` as a dummy variable.

Yathartha22 replied:
Thanks for clarifying this. I will make the changes.

Yathartha22 replied:
> an expression can have symbol as a dummy variable.

@smichr What does it mean? Even if the expression has a dummy variable we need to solve it, right?

Also I don't get how `foo.has(symbol)` is a weaker assertion `symbol in foo.free_symbols`, they seem to work the same when expression has dummy variable as `symbol`

```python
>>> x = Dummy('x')
>>> f = 3*x
>>> f.has(x)
True
>>> x in f.free_symbols
True
>>> y = symbols('x')
>>> f.has(y)
False
>>> y in f.free_symbols
False
```

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,365 @@ def _solveset(f, symbol, domain, _check=False):
+    This form can be easily handed by `solveset`.
+    """
+
+    try:
```
Could you make the assertion more direct? It is not transparently obvioius that `ordered` won't raise a ValueError for some reason. How about:

```
if not (isinstance(f, Add) and len(f.args) == 2:
    # solving for the sum of more than two powers is possible
    # but not yet implemented
    return None
```

Yathartha22 replied:
Yeah I thought too about this, but assumed that we have stated in the documentation so it would be obvious. I guess we can do it as you said.

smichr replied:
trust (that the user read documentation) and verify (that they obeyed the rules) :-)

Yathartha22 commented:
Apart for this @smichr does this PR need more changes, or does it looks good.

Yathartha22 committed:
improved case for handling more than 2 terms in `expo_solver`

Yathartha22 committed:
minor documentation changes

Yathartha22 committed:
removed unnecessay lines from documentation

Yathartha22 commented:
ping @aktech

smichr committed:
do test before working with args

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,344 @@ def _solveset(f, symbol, domain, _check=False):
+
+    This form can be easily handed by `solveset`.
+    """
+    from sympy.core import Add
```
did you try a top import (same line as the one importing Mul)?

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,344 @@ def _solveset(f, symbol, domain, _check=False):
+    >>> check(cos(2**x), x)
+    False
+    """
+    from sympy.core import Add, Mul
```
use top import

Yathartha22 committed:
added test for `expo_solver` to return `None`; used top import for Add and Mul

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    How `\_transolve` is better than `\_tsolve`
+    ===========================================
+
+    1) Better output
```
I would put this in the commit note. Once tsolve is gone this will have no significance.

Yathartha22 replied:
> Once tsolve is gone this will have no significance.

Yeah you are right but I guess until `transolve` completely overtakes `tsolve` we need to have comparisons (for developers to know why to use it).

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    The API of `\_transolve` is designed such that it is easily extensible,
+    i.e, it is easy to add a new class of equation solver without having
+    to mess with the API or solving code of other helpers. The idea
+    behind `\_transolve` is making seperate helpers for different equation
```
is making ->is to make
solver -> solvers

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    method implemented directly in the helpers without interfering
+    with the actual structure of the API.
+
+    4) Less complex API
```
again, no need for comparison here

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    API structure is a mess with lots of recursive and function
+    calls making it hard to understand.
+
+    5) Faster Computation
```
remove comparison (also, every class of equation that transolve attempts to identify will slow *it* down, too)

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    i.e, it is easy to add a new class of equation solver without having
+    to mess with the API or solving code of other helpers. The idea
+    behind `\_transolve` is making seperate helpers for different equation
+    solver. So to make new class of equation solve, you must add a
```
So to... -> So to solve a given class of equation, one helper identifies the equation and another solves the equation or modifies it into a form that can be solved.

Actually, I don't think the helper is solving the equation, it is just recasting the equation into a tractable form, isn't it?

Yathartha22 replied:
> Actually, I don't think the helper is solving the equation, it is just recasting the equation into a tractable form, isn't it?

The thought is helpers can do either thing, convert into a tractable form or solve completely. Like
`expo_solver` reduces the equation whereas `log solver` would return the solution (atleast as of now)

smichr replied:
Where is log solver? 

Would it make sense to name the helpers like those that are already present in solveset: `_solve_abs`, `_solve_trig`, etc...?

smichr replied:
Is `_solve_transcendental` really extensible? What other sorts of equations besides 2 (or more) exponentials will it solve? Should it maybe just be `_solve_exp`?

Yathartha22 replied:
> Would it make sense to name the helpers like those that are already present in solveset: _solve_abs, _solve_trig, etc...?

Yes this can be done. Thanks!

Yathartha22 replied:
> Is _solve_transcendental really extensible? What other sorts of equations besides 2 (or more) exponentials will it solve? Should it maybe just be _solve_exp?

Other sorts of equations will include, logarithmic equations, lambert type equations and maybe we can incorporate solving by `solve_decomposition`.

Yathartha22 replied:
Yeah, this sounds better. I will keep this in mind while implementing other solvers.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    >>> from sympy.solvers.solveset import _transolve as transolve
+    >>> from sympy.solvers.solvers import _tsolve as tsolve
+    >>> from sympy import symbols, S, pprint
+    >>> x = symbols('x', real=True)
```
In keeping with solveset, the results should not depend on the symbol assumptions. If you need to make the symbol real, you can do so internally.

smichr replied:
This is the danger of the docstring example, I suppose: making it look like transolve can/should be called independently of solveset. Perhaps a note clarifying this interaction would be in order. (Or indicate at the outset of the examples that although transolve is being called directly here, normally it would be called by solveset that would take care of putting the expression in a canonical form and passing the relevant assumptions on the symbols.

(But if assumptions are passed on the symbol, then why is the domain necessary? Perhaps any domain clipping should be done back in solveset. But take this comment noting that I am not yet thoroughly familiar with the workings of transolve.)

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+
+    `\_transolve` should be used as an internal function. There should be
+    an internal call to `\_transolve`, passing all the three above
+    parameters as arguments. These three arguments should be sympy
```
sympy -> SymPy

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+
+    * Philosophy behind the module
+
+    `\_transolve` comes into action when `solveset` is unable to solve
```
```
>>> transolve((2**x - 4**x)/(3**x),x,S.Reals)
ConditionSet(x, Eq(3**(-x)*(2**x - 4**x), 0), S.Reals)
>>> solveset((2**x - 4**x)/(3**x),x,S.Reals)
{0}
```

So who is resonsible for getting the equation into a given form?

Yathartha22 replied:
When the equation is passed via `solveset` , it breaks down to `2**x - 4**x` and `3**(-x)` and the former is easily handled by `transolve`, but equation passed directly in transolve would not yield the result as it does not break the equation.

We can have an `lhs.is_Mul` case that will invoke `_solveset` to handle the equation.

Yathartha22 replied:
> I would raise a NotImplementedError and say something like "This Mul expression should be handled in solveset."

Will raising a `TypeError` or `ValueError` be better because `NotImplementedError` would in a way suggest that the equation can be solved (but this cannot be solved by `transolve` anytime in future).
And yes I would rather add in the docstring that what forms does the `transolve` expect the equation to be.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    should be implemented for generalised cases and should return
+    either `True` if the given equation belongs to the class otherwise
+    `False`. `Solving helpers` needs to be implemented with
+    heuristics or algorithms that solves majority of the equations
```
solves majority --> solve a majority

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+      Name of the `identification helper` should be like  `\_check\_class`
+      where `class` will be the name or abbreviation of the class of
+      equation. The `solving helper` will be named as `\_class\_solver`.
+      For example: for exponential equation it becomes `\_check\_expo`
```
equation -> equations

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    exact solution or a reduced form of the equation which `solveset`
+    can handle.
+
+    Apart from this, few other things needs to be taken care while adding
```
few->a few

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+      For example: for exponential equation it becomes `\_check\_expo`
+      and `\_expo\_solver`.
+    - Be sure to consider corner cases.
+    - Adding tests for each helper.
```
Adding -> Add

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+      - What is the purpose of the helper.
+      - How it solves the equation and how does it identifies the equation.
+      - Examples should be included with its proof of correctness.
+      - What are the input parameters and what does the helper returns.
```
shouldn't the input parameters always be the same for all helpers?

Yathartha22 replied:
I am not sure as of now about it. By doing this we are in a way restricting helper solvers to have a limited number of parameter(s), for eg:  [_check_log](https://github.com/sympy/sympy/pull/14792/files#diff-eec0422923e8f100745c015cd8fdd6cfR1193)
won't require a second parameter `symbol` unlike `_check_expo` therefore only the `f` is passed.

Do we need to restrict input parameters?

smichr replied:
Every check function should get f and symbol and should only be working with the parts of the equation that contain symbol. A function can choose to ignore the symbol (as log currently does) but I would recommend against that: make it separarate args into those that depend on symbol and those that don't and only transform th esymbol-dependent parts. (Otherwise, you can run into empty recursion situations where, for example, log puts things together which then get taken apart.)

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    - Adding tests for each helper.
+    - Adding a docstring to your helper that describes the method
+      implemented.
+      Following things needs to be included while writing the
```
Following -> The following

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    r"""
+    Helper function for solving (supported) exponential equations.
+
+    Exponential equations are a sum of (currently) at most
```
`Exponential equations which are the sum of at most`

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+def _transolve(f, symbol, domain):
+    r"""
+    Function to solve transcendental equations. It is a helper to
+    `solveset` and should be used internally as of now. `\_transolve`
```
I believe it would always be used internally.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    Parameters
+    ==========
+
+    f : can be any transcendental equation that needs to be solved.
```
`can be` -> `is`

smichr commented on sympy/solvers/solveset.py:
> The thought is helpers can do either thing, convert into a tractable form or solve completely

I commented on the log solver PR. IMO you *do not* want to go down this road since this will lead to decentralization. Keep the checking and domain clipping all in one place. Consider making transolve do the following:
1. step through an identify/transform sequence
2. if no identifications were a hit return the original equation or a ConditionSet?
3. if an identification is made return the changed form
4. if the call to transolve gets the same equation back it knows it wasn't a transcendental; if it is changed, `_solveset` passes the equation back to `_solveset` again.

...something like that.

smichr commented:
Here are some tests to consider:

```
>>> solveset(exp(x)-exp(2*(x+1)))
Union({nan}, ImageSet(Lambda(_n, 2*_n*I*pi - 2), S.Integers))  # ans (except NaN should not be there)
>>> solveset(log(exp(x))-log(exp(2*(x+1))))
ConditionSet(x, Eq(log(exp(x)) - log(exp(2*x + 2)), 0), Complexes(S.Reals x S.Reals, False)) # should be ans
>>> solveset(log(exp(x))-((2*(x+1))))
ConditionSet(x, Eq(-2*x + log(exp(x)) - 2, 0), Complexes(S.Reals x S.Reals, False)) # should be -2
>>> solveset(((x))-log(exp(2*(x+1))))
ConditionSet(x, Eq(x - log(exp(2*x + 2)), 0), Complexes(S.Reals x S.Reals, False))  # should be -2
```

Yathartha22 committed:
removed `pow_type()` as this case is handled in `solveset`; removed loop in `transolve`; modified documentation and tests;

Yathartha22 commented:
Major things addressed in the latest commit:

- I have removed `pow_type()` because this wasn't needed because such equations will be handled in `_invert` itself.

- Also removed the loop in `transolve` as it is not required because `transolve` will receive only those equations that will have only one argument in the `rhs_s` (see its invocation).

- removed condition `lhs == symbol` as this is too redundant (it is already handled in `solveset`).

- improved documentation of `How to use transolve` as well as the `Example`.

- removed a test from `test_solve_only_exp_2` and added it in `test_exponential_symbols`. Also shifted the test below so that it is easily seen.

Few doubts:

- Currently the design of `transolve` is such that it cannot be used as an independent function (like `_tsolve` can). Cases like `2**x - 32`, `(2**x - 4**x)/4**(x)` etc. will not be handled by `transolve` independently. The reason being that if handling is done for these cases it would cause unnecessary repetition of code.
Is this thing okay for `transolve` ?

- If yes what should be returned if the above-like equations are passed directly to `transolve`. Currently they would return a `ConditionSet`.
And I think adding a case for raising any kind of error for such equations (as mentioned in [this comment](https://github.com/sympy/sympy/pull/14736#discussion_r197514203)) won't be good enough as we will need to add many cases and also we will not be sure whether the equations will actually be handled or not by `solveset`.
I guess `ConditionSet`  will be okay as of now (as ConditionSet instance means in solveset the equation cannot be handled), I am not sure though.

smichr commented:
> Is this thing okay for transolve ?

`transolve` should really be `_transolve`; nobody should be calling `transolve` directly. It is an "implementation detail" of `solveset`. Alternatively, if not already present, it could not be listed in public facing function from the file...`__files__ = [blah blah blah]` (or something like that) wouldn't have `transolve` listed. I prefer making it explicitly private with the leading underscore.

Yathartha22 commented:
@smichr `transolve` is really `_transolve` (Sorry I have been using the name `transolve` instead of `_transolve`).
I just wanted it to be clear and that is why I have removed conditions like `lhs == symbol` (which I included in previous commits).
So we should not be actually concerned on what `_transolve` returns when called independently, right?. Is it okay to add [examples](https://github.com/sympy/sympy/pull/14736/files#diff-eec0422923e8f100745c015cd8fdd6cfR1117) in the docstring? Also what about [this](https://github.com/sympy/sympy/pull/14736/files#diff-eec0422923e8f100745c015cd8fdd6cfR1107) explanation, is it okay to have it?

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,331 @@ def _solveset(f, symbol, domain, _check=False):
+    >>> from sympy.solvers.solveset import _transolve as transolve
+    >>> from sympy.solvers.solvers import _tsolve as tsolve
+    >>> from sympy import symbols, S, pprint
+    >>> # symbol modified as would have been by solveset
```
When solveset calls transolve it will pass assumptions with
the symbols. For demonstration purposes, we need to do so,
too.

>>> x = symbols('x', real=True)

smichr commented:
> what about this explanation

I don't seem to land on any explanation when I follow this.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,331 @@ def _solveset(f, symbol, domain, _check=False):
+
+    1) Better output
+
+    `\_transolve` provides a better output to some equations than
```
`\_transolve` provides expressions in a more simplified form:

(and delete the rest)

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,331 @@ def _solveset(f, symbol, domain, _check=False):
+         | log(2/9)|
+    [-log\2         /]
+
+    `\_transolve's` output seems more simple, readable and appropriate
```
remove

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,331 @@ def _solveset(f, symbol, domain, _check=False):
+    2) Extensible
+
+    The API of `\_transolve` is designed such that it is easily
+    extensible, i.e, it is easy to add a new class of equation solver
```
i.e. the code that solves a given class of equations is encapsulated in a helper
and not mixed in with the code of `\_transolve` itself.

and remove the rest

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,331 @@ def _solveset(f, symbol, domain, _check=False):
+    equation, one helper identifies the equation and another solves
+    the equation or modifies it into a form that can be solved.
+
+    3) Modular
```
remove 3

Yathartha22 replied:
I am not sure about it because while designing `_transolve`, modularity is taken care of and this is what makes it different than `_tsolve`

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,331 @@ def _solveset(f, symbol, domain, _check=False):
+    4) Less complex API
+
+    The API of `\_transolve` is designed in a way that it is less
+    complex to read and understand unlike in `\_tsolve` where the
```
It sounds like you are talking about the code rather than the API. I really doubt transolve is going to solve an equation with less recursion than tsolve does. It's just that any modification steps are located outside of transolve, isn't it?

Yathartha22 replied:
> I really doubt transolve is going to solve an equation with less recursion than tsolve does

Yes this could be, still `transolve` solves equation faster than `_tsolve`

```python
>>> import time
>>> f = 3**(2*x) - 2**(x + 3)

>>> start_time = time.time()
>>> _tsolve(f, x)
>>> print("--- %s seconds ---" % (time.time() - start_time))
--- 0.4206206798553467 seconds ---
>>> start_time = time.time()
>>> _transolve(f, x, S.Reals)
>>> print("--- %s seconds ---" % (time.time() - start_time))
--- 0.05263113975524902 seconds ---
```
One reason is probably a lot of function calls.

> It sounds like you are talking about the code rather than the API

I am in the impression that when we talk about API, we are indirectly talking about the code, because code is what API is built with. Am I getting this wrong?

> It's just that any modification steps are located outside of transolve, isn't it?

Yes.

smichr replied:
> Am I getting this wrong?

In the context of SymPy I would think of the API as the information (list of arguments) that are required for calling a function. What the function does between receiving the request and returnin a result is not the API (as I, not a profefssional in this field, have come to understand the term). So take that fo what it's worth.

Yathartha22 replied:
I guess we should remove this point.

Yathartha22 commented:
> I don't seem to land on any explanation when I follow this.

It suggests users that how `_transolve` is/can be used: it is an internal function and takes input params as passed from solveset and there are chances that these params get modified (like the assumptions will be passed in the `symbol` and may be only a part of the equation is taken care as in `(2**x - 4**x)/3**x`) and to use it as an independent function we require to pass arguments in similar way as `solveset` passes.
Do I need to change/rephrase it?

smichr commented:
> raising any kind of error for such equations

You need to know what you expect to handle in tsolve. If you can handle anything with the help of solveset then transolve is not really a helper for solveset, solveset is a helper for transolve. The purpose of raising an error for an unexpected form is to force the calling program to send the expected form of an expression so you don't have to do all (or repeat much) of the work there.

smichr commented:
Until you wire transolve up to solveset, it won't be clear what your expectations are for the expressions received from solveset. Also, if you call solveset (or variant) from transolve will it do error checking? And then will the error checking be repeated once the solution from transolve is returned?

Yathartha22 committed:
minor document changes

Yathartha22 commented:
> Until you wire transolve up to solveset, it won't be clear what your expectations are for the expressions received from solveset.

I am not sure but should we leave it for later? because handling error would unnecessarily make things complicated. We should let `_transolve` return `ConditionSet` for such equations as of now.

> Also, if you call solveset (or variant) from transolve will it do error checking?

I don't think so. But there could be a case when solveset calls transolve again and if at this point error is raised solutions won't be returned. So I guess raising error is risky.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+    if not (isinstance(f, Add) and len(f.args) == 2):
+        # solving for the sum of more than two powers is possible
+        # but not yet implemented
+        return None
```
The function doc string doesn't mention that it can return `None` as well.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+
+    * Proof of correctness of the method
+
+    The logarithm is the inverse function to exponentiation.
```
`The logarithm function is the inverse of the exponential function`

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+    Therefore if we are given an equation with exponent terms, we can
+    convert every term to its corresponding log form. This is achieved by
+    taking logarithms and expanding the equation using log identities
+    so that it can be easily handled by `solveset`.
```
so that it can easily be handled by `solveset`.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+
+    for add_arg in Add.make_args(f):
+        for mul_arg in Mul.make_args(add_arg):
+            if isinstance(mul_arg, (Pow, exp)) and \
```
avoid `\` rather use parenthesis, for e.g.:

```
if isinstance(mul_arg, (Pow, exp)) and (
                   symbol in mul_arg.exp.free_symbols):
```

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+    Solving helpers: These are the helpers that solve the equation.
+    They are invoked once the equation is identified by the `identifying
+    helpers`. These functions return either the complete solution or
+    a modified version of the equation that `solveset` might better
```
better -> be able to

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+
+    `\_transolve` comes into action when `solveset` is unable to solve
+    the equation as a last resort to get the solutions. The idea is to
+    get the equation in a case (Add, Pow etc.) that satisfies its
```
rephrase this to be more meaningful: *get the equation in a case*

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+    `a*f(x) + b*g(x)`, where `f(x)` and `g(x)` are power terms,
+    therefore they are included in `Add` case.
+
+    Once done different `identifying helper` conditions are included
```
rephrase this as well: _Once done different_

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+    therefore they are included in `Add` case.
+
+    Once done different `identifying helper` conditions are included
+    to check which class the equation belongs. `Solving helper` is
```
belongs to

aktech replied:
solving helper should be italic instead of in backticks

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+    def add_type(eq, symbol, domain):
+        """
+        Helper for `_transolve` to handle equations of
+        `Add` type.
```
Add a trivial example of add type

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+        """
+        result = ConditionSet(symbol, Eq(eq, 0), domain)
+        # try factoring it;
+        # powdenest is used to try get powers in standard form
```
try to

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+        return None
+
+    lhs, rhs = list(ordered(f.args))
+
```
remove empty line, (as it is not separating a big enough block of code)

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+    lhs, rhs_s = invert_complex(f, 0, symbol, domain)
+
+    if isinstance(rhs_s, FiniteSet):
+        rhs = rhs_s.args[0]
```
I added loop for this case before but now it seems to me that loop isn't necessary because `_solveset` already uses loop [here](https://github.com/sympy/sympy/blob/master/sympy/solvers/solveset.py#L935) and `_transolve` is called within this loop, therfore `rhs_s` will always have one argument

aktech replied:
That's fine, but add an assert here to make your intentions clear, so that if someone changes that in `solveset` tomorrow, transolve should tell the world that it broke.

Yathartha22 commented:
Adding to [this](https://github.com/sympy/sympy/pull/14736#discussion_r198129763) comment, we should have the API of all the `identification helpers` (check_log, check_expo) same and that of `solving helpers` can vary. Is this okay?

Yathartha22 committed:
documentation imporovements

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,318 @@ def _solveset(f, symbol, domain, _check=False):
+      `\_solve\_class`.
+      For example: for exponential equations it becomes `\_check\_expo`
+      and `\_solve\_expo`.
+    - *Identification* *helpers* should take two input parameters,
```
Is this point okay for the API design of the helpers?
Ref: [this](https://github.com/sympy/sympy/pull/14736#discussion_r198129763) comment

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,318 @@ def _solveset(f, symbol, domain, _check=False):
+
+    Solving helpers: These are the helpers that solve the equation.
+    They are invoked once the equation is identified by the `identifying
+    helpers`. These functions return either the complete solution or
```
If the solving helper only takes as input the function, how will it be able to return a solution? (c.f. line 1239 below)

Yathartha22 replied:
They won't be able to return a solution, I would rather remove "complete solution". The idea now is that the solving helpers will convert the equation into a tractable form and then in the `_transolve` itself, the changed equation can be handled. This is how `_solve_expo` works and we will make the other helpers work in similar fashion. For this scenario I guess having the `equation (f)` as the only argument for `solving helpers` would work?

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,318 @@ def _solveset(f, symbol, domain, _check=False):
+
+    `\_transolve` comes into action when `solveset` is unable to solve
+    the equation as a last resort to get the solutions. The idea is that
+    the equation is determined for its case (Add, Pow etc.) that satisfies
```
These two paragraphs are confusing as written. What about this (or something like it)?

The purpose of `_transolve` is to take equations which are not already polynomial in their generator(s) and to either recast them as such through a valid transformation or to solve them outright. A pair of helper functions for each class of supported transcendental functions are employed for this purpose. One identifies the transcendental form of an equation and the other either solves it or recasts it into a tractable for that can be solved by solveset. For example, an equation in the form `a*b**f(x) - c*d**g(x) = 0` can be transformed to `log(a) + f(x)*log(b) - log(c) - g(x)*log(d) = 0` (under certain assumptions) and this can be solved with solveset if `f(x)` and `g(x)` are polynomial in form.

Yathartha22 replied:
This is great @smichr looks better than the previous one. Thanks!

Yathartha22 replied:
> ... and to either recast them as such through a valid transformation or to solve them outright.

Solving them outright would mean `_transolve` completely solves it, but it just modifies the equation, solving will be done via `_solveset`. I should remove " solve them outright."

Yathartha22 commented:
Well, I see a problem if we make the API of the solving helpers return only modified form for the equation (by using only `f` as the parameter). While exponents and log are in accordance to it, but when implementing Lambert type equations we will take help of the bivariate module which will return the exact solution at once. Therefore we would need `symbol` argument for solving helpers.
But what should we do for the ones that won't return exact solutions (for these `symbol`  argument is not needed)??

smichr commented:
> But what should we do for the ones that won't return exact solutions

Provide it...and ignore it. If the routine chooses to ignore it and just return a modified equation, that is fine. 

But I'm still not convinced that just calling the identifiers and helpers from solve is not the way to go:

```python
def solveset(eq, ...):
    while True:
        if eq.is_Mul:
            return union of solutions of factors
        # eq is Add, Pow, Function, or Expr like Sum, Integral, etc...
        identify gens
        if polynomial in single gen:
            return polynomial solution after mapping generator inversion on it
        elif _is_exponential(eq, x):
            rewrite eq
        elif _is_logorithmic(eq, x):
            return solve as log
        elif _is_trigonometric(eq, x):
            rewrite eq
        elif _is_lambert(eq, x):
            return lambert solutions
        else:
            raise NotImplementedError
```
(gross simplifcation, but that's the general idea) Why do we need to go to `transolve` to do this?

As I recall, however, for the lambert case there are many forms it can take and I thought it better to do the solving immediately after doing the identification. It seems that you will throw away all the information used in identification when you pass the equation to the solver.

Yathartha22 commented:
The intention of creating `_transolve` is to separate solving of transcendental equations from the `solveset` otherwise it can create a mess in the code of solveset (while adding something new or while debugging). We have been focusing on modularity and extensibility throughout while designing `_transolve`

And I guess the current flow is somewhat similar to what you are trying to suggest; rather than checking different transcendental types in `solveset` itself it calls for a seperate helper (transolve) and there it checks different classes of transcendental equations.

```python

def solveset(eq, ...):
    while True:
        if eq.is_Mul:
            return union of solutions of factors
        # eq is Add, Pow, Function, or Expr like Sum, Integral, etc...
        identify gens
        if polynomial in single gen:
            return polynomial solution after mapping generator inversion on it
        elif _is_trigonometric(eq, x):
            rewrite eq
        elif _is_transcendental(eq, x):
             _transolve(eq, x)
        else:
            raise NotImplementedError

def _transolve(eq,..):
   if eq.is_exponential(eq, x):
       _solve_expo(eq, x) # to rewrite eq 
   elif eq._is_logarithmic(eq, x):
       _solve_log(eq, x)  # to rewrite eq
   elif eq._is_lambert(eq, x):
       _solve_lambert(eq, x)   # return lambert solutions
```
The intention is to have modularity at every step so that if changes are to be made further, only that particular function is being modified without disturbing the existing code.
Hope the idea and the implementation convinces you :)

Yathartha22 committed:
improved `Philosophy behind the module`; changed the naming convention of identification helpers; minor changes in `add_type()`

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1136,24 +1136,24 @@ def _transolve(f, symbol, domain):
 
     Solving helpers: These are the helpers that solve the equation.
```
-> Solving helpers: once an equation is identified, a corresponding helper either solves the equation or returns a form of the equation that `sovleset` might better be able to handle.

Yathartha22 committed:
minor documnetation changes

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1201,24 +1200,21 @@ def _transolve(f, symbol, domain):
+    The design of `\_transolve` makes adding a new class of equation
+    solver easy.
+
+    Decide from where the *identification* *helper* will be invoked.
```
Is there any transcendental equation (other than something like cos(x) = 0) that can be solved/handled by transolve that is not an Add? 

This paragraph is confusing with the example put in the middle of the instructions. "Once done" and "add a call" are not clearly defined. It would be better to give a code snippet like

```python
if eq.is_Add:
    ...
    if _is_expo(eq, x):
        new_eq = _solve_expo(eq, x)
    ...
```

Also, what to name the result should probably be specified. Are different names used for the case when a solution is returned vs the case where a rewritten form is returned?

Yathartha22 replied:
> Is there any transcendental equation (other than something like cos(x) = 0) that can be solved/handled by transolve that is not an Add?

I can't think of any other than `Add` as of now if there are we can create another internal function for it.

>It would be better to give a code snippet like

I will add the code snippet for better understanding.

> Are different names used for the case when a solution is returned vs the case where a rewritten form is returned?

Yes, when the equation is recasted then `new_eq` will be updated and then `solveset` will be called to update the `result`, whereas when solution is returned the `result` is directly updated

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -1201,24 +1200,21 @@ def _transolve(f, symbol, domain):
+    invocation condition is placed inside `Add` case.
+    Once done, add a call to its corresponding *solving* *helper*.
+
+    For the class of equation define *identification* and
```
The previous paragraph and this one need to be edited. The last one talks about th esolving helper but this paragraph sound like *it* is trying to introduce the helpers. I think there are three steps (if there are more than Add-type transcendentals):
1. identify class
2. write identifier
3. write solvers

Yathartha22 replied:
Okay explaining using three steps would be better:

```
Adding a new class of equation solver is a three-step procedure:

- Identify the type of the equations

Determine the type of the class of equations to which they belong, it could be of `Add`, `Pow`, 

etc. types. Seperate internal functions are used for each type, if it is already present include

identification and solving helpers within the function otherwise add a new internal function and then 

include identification and solving helpers within it.

### code snippet can be added here

- Define the identifying helpers

- Define the solving helpers
```

Something like this is okay?

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,310 @@ def _solveset(f, symbol, domain, _check=False):
+    lhs, rhs = list(ordered(f.args))
+    log_type_equation = expand_log(log(lhs)) - expand_log(log(-rhs))
+
+    return log_type_equation
```
What is going to happen when the user selected the complex domain for the solution in solveset? The symbol won't be real, will it? And then the equation won't simplify as desired. I think you will get an infinite loop. But if there is no solution with numbers having an imaginary part then the real solution should be returned...and somewhere that change to a real variable will have to take place. I suspect that when solving in the complex domain an attempt at the complex solution will have to be followed with an attempt at a real solution.

Yathartha22 replied:
> What is going to happen when the user selected the complex domain for the solution in solveset?

The equation won't get simplified (by `expand_log`) but still it will be transformed to `log` hence it won't fall into infinite loop.

As of now untill `_transolve` gets finalised, @aktech suggested to not focus on solving in complex domain. I guess once this PR gets into a proper shape, I can add solving in complex domain in a seperate PR. What do you say?

aktech replied:
Not solving in complex domain is fine, but it should not break with an infinite loop.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -943,7 +945,12 @@ def _solveset(f, symbol, domain, _check=False):
+                        result_rational = _solve_as_rational(equation, symbol, domain)
+                        if isinstance(result_rational, ConditionSet):
+                            # may be a transcendental type equation
+                            result += _transolve(equation, symbol, domain)
```
If you aren't going to implement solving in the complex domain then you should add the test here, perhaps:

```python
...
# may be a transcendental type equation
if not domain.is_subset(S.Reals):
    raise NotImplementedError('solving transcendent equation in the complex domain.')
result += _transolve(...
```

Yathartha22 replied:
Is getting a `ConditionSet` not a good idea?
As of now for the complex domain, `ConditionSet` is returned.

smichr replied:
oh, sorry. ConditionSet is fine.

Yathartha22 committed:
Improved documentation for "How to add new class of equations"

Yathartha22 committed:
added `make_expr_args()` to get all the terms of the expressions

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,330 @@ def _solveset(f, symbol, domain, _check=False):
     return result
 
 
+def make_expr_args(f):
```
As of now the result of the routine will be used by `identifying helpers`.

smichr replied:
I suggest naming this `term_factors` since it returns the factors of all terms (contrary to what the current docstring says).

Yathartha22 replied:
I don't think it returns factors of the expression, it just gives the terms as it is present in the expression

```python
>>> make_expr_args(x**2 - 1)
[-1, x**2]
```

Yathartha22 replied:
Also will it be a good idea to have this method placed somewhere else, this could be used as a public function (I don't think sympy has anything of this sought).

Yathartha22 replied:
Oh sorry :), I miss understood a bit. I will update the name as well as the docstring

Yathartha22 committed:
renamed `make_expr_args` to `term_factors()` and improved its documentation

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -985,9 +985,10 @@ def _solveset(f, symbol, domain, _check=False):
 
 
-def make_expr_args(f):
+def term_factors(f):
```
Sounds good @smichr. I will make changes.
Is this the best place for this kind of a function?

Yathartha22 committed:
made `term_factors()` an iterator; added some examples in its docstring

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -987,14 +987,25 @@ def _solveset(f, symbol, domain, _check=False):
+    >>> from sympy import symbols
+    >>> from sympy.solvers.solveset import term_factors
+    >>> x = symbols('x')
+    >>> term_factors_obj = term_factors(x**2 - 1)
```
```
>>> list(term_factors(-2 - x**2 + x*(x + 1)))
[-2, -1, x**2, x, x + 1]
```

Yathartha22 replied:
Should I remove the loop? It will probably demonstrate another way to use the results from the function rather than converting into a list or tuple

Yathartha22 committed:
removed unused list in `term_factors()`; improved doctests for the same

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,338 @@ def _solveset(f, symbol, domain, _check=False):
+    can be transformed to
+    `log(a) + f(x)*log(b) - log(c) - g(x)*log(d) = 0`
+    (under certain assumptions) and this can be solved with `solveset`
+    if `f(x)` and `g(x)` are polynomial in form.
```
*are in polynomial form.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -978,6 +985,338 @@ def _solveset(f, symbol, domain, _check=False):
     return result
 
 
+def term_factors(f):
```
please make it private

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -10,14 +10,17 @@
 from sympy.core.sympify import sympify
-from sympy.core import S, Pow, Dummy, pi, Expr, Wild, Mul, Equality
+from sympy.core import (S, Pow, Dummy, pi, Expr, Wild, Mul, Equality,
+                        Add, Mul)
```
`Mul` imported twice

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -10,14 +10,17 @@
 from sympy.core.facts import InconsistentAssumptions
 from sympy.core.numbers import I, Number, Rational, oo
-from sympy.core.function import (Lambda, expand_complex, AppliedUndef, Function)
+from sympy.core.function import (Lambda, expand_complex, AppliedUndef, Function,
```
Please remove unused import `Function`

aktech reviewed (commented):
For all the function that, you have created irrespective of private or public, add the following type of params and return type (for consistency):

```
    Parameters
    ==========

    f : Expr or a relational.
        The target equation or inequality
    symbol : Symbol
        The variable for which the equation is solved
    domain : Set
        The domain over which the equation is solved

    Returns
    =======

    Set
        A set of values for `symbol` for which `f` is True or is equal to
        zero. An `EmptySet` is returned if `f` is False or nonzero.
        A `ConditionSet` is returned as unsolved object if algorithms
        to evaluate complete solution are not yet implemented.

    `solveset` claims to be complete in the solution set that it returns.

    Raises
    ======

    NotImplementedError
        The algorithms to solve inequalities in complex domain  are
        not yet implemented.
    ValueError
        The input is not valid.
    RuntimeError
        It is a bug, please report to the github issue tracker.

```

aktech commented:
Update the docstring at the top of the file `solveset.py` to include the solving of transcendental equations.

aktech commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1749,3 +1730,187 @@ def test_issue_14454():
+
+
+@XFAIL
+def test_uselogcombine_2():
```
Instead of 1 or 2 as function name suffix, write the difference between them as function name.

Yathartha22 replied:
I don't see a difference here, these tests were already present I just moved them here.
I would rather include all in one.

Yathartha22 committed:
made `term_factors()` private; minor documentation changes; renamed log tests function name

Yathartha22 committed:
removed repeated test and minor change in documentation

smichr commented:
The only question I had about design was whether there is any transcendental equation that is not of Add type that this solver will be attempting to solve? (I though I asked this but don't see it in the comments that are still showing.)

Yathartha22 commented:
> The only question I had about design was whether there is any transcendental equation that is not of Add type that this solver will be attempting to solve?

`_transolve` will be seeing the lhs of the equation of being Add, Mul, or Pow type. Currently, I have included only the add type because only solving exponential equations is implemented (which is basically of the add type).
There are `Pow` type of lhs like `x**x - 2 = 0` (which can be solved in lambert form). So we need to include Pow case while implementing the lambert solver.
If there are any other Mul type of lhs we can add a different helper and solve it accordingly, the thing is for this particular PR there doesn't seems a case, but needs to be taken care while implementing logarithmic and lambert solvers

smichr commented:
> seeing the lhs of the equation

thanks for recap. +1

aktech reviewed (approved):
Looks good to me.

aktech commented:
@Yathartha22 thanks for the work! @smichr thanks for reviewing! This is in.
